### ISSUE TITLE
efivar 26: errors when trying to build from source

### ISSUE THREAD

**Issue by tgurr** [rhinstaller/efivar#60]
My compile params are:

```
bindir=/usr/$(exhost --target)/bin
includedir=/usr/$(exhost --target)/include
libdir=/usr/$(exhost --target)/lib
prefix=/usr/$(exhost --target)
CROSS_COMPILE=$(exhost --tool-prefix)
BINTARGETS=efivar
STATICLIBTARGETS=
```

When trying to build efivar 26 I get the following error:

```
x86_64-pc-linux-gnu-cc -pipe -O2 -march=native -I/var/tmp/paludis/build/sys-boot-efivar-26/work/efivar-26/src/include/efivar/    -L.      -pipe -O2 -march=native -shared \
  -Wl,-soname,libefivar.so.26 \
  -Wl,--version-script=libefivar.map \
  -o libefivar.so dp.o dp-acpi.o dp-hw.o dp-media.o dp-message.o efivarfs.o error.o export.o guid.o guids.o guid-symbols.o lib.o vars.o -ldl 
make[1]: Leaving directory '/var/tmp/paludis/build/sys-boot-efivar-26/work/efivar-26/src'
lib.o:(*IND*+0x0): multiple definition of `efi_set_variable'
lib.o:lib.c:(.text+0xf0): first defined here
collect2: error: ld returned 1 exit status
make[1]: *** [/var/tmp/paludis/build/sys-boot-efivar-26/work/efivar-26/Make.rules:20: libefivar.so] Error 1
make: *** [Makefile:11: all] Error 2
```

When I apply a patch found at openembedded (source: https://github.com/openembedded/meta-openembedded/blob/master/meta-oe/recipes-extended/efivar/efivar/allow-multi-definitions-for-native.patch ) I get the following error:

```
x86_64-pc-linux-gnu-cc -pipe -O2 -march=native -I/var/tmp/paludis/build/sys-boot-efivar-26/work/efivar-26/src/include/efivar/    -L.      -pipe -O2 -march=native -shared \
  -Wl,-soname,libefiboot.so.26 \
  -Wl,--version-script=libefiboot.map \
  -Wl,-z,muldefs \
  -o libefiboot.so crc32.o creator.o disk.o gpt.o linux.o loadopt.o -lefivar 
x86_64-pc-linux-gnu-cc -pipe -O2 -march=native -I/var/tmp/paludis/build/sys-boot-efivar-26/work/efivar-26/src/include/efivar/    -L.      -pipe -O2 -march=native -o efivar efivar.c -lefivar -ldl -lpopt

Error:
  * In program cave perform install --hooks --managed-output --output-exclusivity with-others =sys-boot/efivar-26:0::arbor --destination installed --x-of-y 1 of 1:
  * When installing 'sys-boot/efivar-26:0::arbor':
  * When running an ebuild command on 'sys-boot/efivar-26:0::arbor':
  * Install failed for 'sys-boot/efivar-26:0::arbor' (paludis::ActionFailedError)

make[1]: Leaving directory '/var/tmp/paludis/build/sys-boot-efivar-26/work/efivar-26/src'
./libefivar.so: undefined reference to `strdupa'
collect2: error: ld returned 1 exit status
make[1]: *** [/var/tmp/paludis/build/sys-boot-efivar-26/work/efivar-26/Make.rules:13: efivar] Error 1
make: *** [Makefile:11: all] Error 2
```

**Comment by vathpela** [comment]
AFAICS this first error comes from building part of the tree with -flto and part without it.  If you do "make clean" it should fix itself?

As for strdupa, I have no idea what's going on unless you're linking against a non-standard glibc or somehow your cross compiler headers are screwed up.

**Comment by superm1** [comment]
I was getting this when building efivar 30 for Debian too.  Adding -flto to CFLAGS of the build process does fix it.

**Comment by tgurr** [comment]
make clean doesn't help since the package manager always starts a build with a clean environment, either the fix mentioned above or adding -flto to CFLAGS like @superm1 wrote is a workaround for the first error.

However I'm still running into the strdupa error when trying to build efivar 30. I'm building with glibc 2.24 installed and I didn't run into similar issues yet with other packages, so I guess my headers should be okay.

**Comment by superm1** [comment]
Adding -flto to CFLAGS seemed to fix it for Debian on relevant architectures, but Ubuntu is failing on armhf and arm64 still even with that.

armhf:
https://launchpadlibrarian.net/290320991/buildlog_ubuntu-zesty-armhf.efivar_30-1_BUILDING.txt.gz

arm64:
https://launchpadlibrarian.net/290320987/buildlog_ubuntu-zesty-arm64.efivar_30-1_BUILDING.txt.gz

**Comment by vathpela** [comment]
https://github.com/rhinstaller/efivar/pull/70 should also fix this case.

**Comment by tgurr** [comment]
Patch from #70 works for me. Still running into the 

> undefined reference to `strdupa'

 error after though.

**Comment by judu** [comment]
@tgurr I had the same issues.
FWIW, I did that to fix the strdupa error: [0002-Fix-flags-to-enable-strdupa.txt](https://github.com/rhinstaller/efivar/files/721369/0002-Fix-flags-to-enable-strdupa.txt).

The trick is to enable `_GNU_SOURCE` and `__USE_GNU` so that string.h can properly define `strdupa`. Somehow, when running make manually, it worked, but the exheres seemed to lack something… could not find what.

I might have been a bit heavy on the -D_GNU_SOURCE stuff, but I wasn't sure where to put it exactly. Feel free to pick whatever you want, would you wish to apply it. ;)

**Comment by tgurr** [comment]
@judu Thank you very very much for your help here (and on the Exherbo package!), your fix works fine for me.

**Comment by vathpela** [comment]
I don't understand why you need to define __USE_GNU.  features.h (which string.h includes) just has:
```
#ifdef  _GNU_SOURCE
# define __USE_GNU      1
#endif
```
Which means we should just always get it defined.  What's causing that not to be the case?

**Comment by judu** [comment]
Well, I'm not the expert at all when it comes to that stuff.
I just fumbled around and this `#define` did the trick.
I have no idea why. I found the same stuff in features.h as you did.

As I wrote: Somehow, when running make manually, it worked, but the exheres seemed to lack something… could not find what.

It was like 4 in the morning already, I needed to fix my system boot, so I went for the dirty stuff. ;)

Anyway, I think this is a problem for the exherbo packaging rather than an issue with your program/compilation tools.
